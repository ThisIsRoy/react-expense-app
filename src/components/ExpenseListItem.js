import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


const ExpenseListItem = ({ dispatch, id, description, amount, createdAt, browserHistory }) => (
  <div>
    <h3>{description}</h3>
    <p>Amount: {numeral(amount / 100).format('$0,0.00')}</p>
    <p>Created: {moment(createdAt).format('MMMM Do, YYYY')}</p>
    <Link to={"/edit/" + id}><button>Edit</button></Link>
  </div>
);

export default ExpenseListItem;