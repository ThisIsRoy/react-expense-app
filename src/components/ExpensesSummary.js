import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
  const expensesCount = props.expenses.length;
  const expensesSum = numeral(selectExpensesTotal(props.expenses) / 100).format('$0,0.00');

  if (expensesCount === 0) {
    return (<h1>There are no active expenses!</h1>);
  } else if (expensesCount === 1) {
    return (<h1>Viewing 1 expense totalling {expensesSum}</h1>);
  } else {
    return (<h1>Viewing {expensesCount} expenses totalling {expensesSum}</h1>);
  }
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);
