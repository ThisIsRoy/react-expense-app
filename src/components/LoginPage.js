import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="login-text">A simple and elegant way of <br/>tracking your expenditures</h1>
      <button className="login-button" onClick={startLogin}>Log In</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);