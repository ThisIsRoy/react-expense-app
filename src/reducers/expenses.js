const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(eachExpense => eachExpense.id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((eachExpense) => {
        if (eachExpense.id === action.id) {
          return {
            ...eachExpense,
            ...action.updates //spread operators to replace old information
          }
        } else {
          return eachExpense;
        }
      });
    case 'SET_EXPENSES':
      return action.expenses;
    default:
      return state;
  }
};

export default expensesReducer;