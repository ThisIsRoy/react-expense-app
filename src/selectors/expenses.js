import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((eachExpense) => {
    const createdAtMoment = moment(eachExpense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = eachExpense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    // sort by newest first 
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
      // sort by most expensive first
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};

export default getVisibleExpenses;