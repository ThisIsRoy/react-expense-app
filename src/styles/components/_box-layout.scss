.box-layout {
  align-items: center;
  background: url('/images/bg.jpg');
  background-size: cover;
  display: flex;
  height: 100vh;
  justify-content: center;
  width: 100vw;
}

.box-layout__box {
  // background: fade-out(green, 0.15);
  color: white;
  text-align: center;
}

.login-text {
  animation: fadein 3s ease-in-out;
  //color: #79C753;
  color: white;
  font-size: $l-size;
  font-weight: 400;
  margin-bottom: $xl-size;
}

@keyframes fadein {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}