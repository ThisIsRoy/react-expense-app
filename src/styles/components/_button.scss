.button {
  background: $blue;
  border: none;
  color: white;
  display: inline-block;
  font-size: $font-size-large;
  font-weight: 300;
  line-height: 1;
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--secondary {
  background: #888888;
}

.login-button {
  background-color: #DC4C46;
  border: 1px solid white;
  border-radius: $s-size;
  color: white;
  font-size: $l-size;
  padding: $m-size $l-size;
  text-align: center;
  text-decoration: none;
  transition: 0.2s ease-in-out;
}

.login-button:hover {
  transform: scale(1.1);
}
