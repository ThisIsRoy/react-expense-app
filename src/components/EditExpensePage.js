import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import expenses from '../selectors/expenses';
import { startEditExpense, startRemoveExpense } from '../actions/expense';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onClick}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(eachExpense => eachExpense.id === props.match.params.id)
});

// removeExpense needs props as an input?
const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);