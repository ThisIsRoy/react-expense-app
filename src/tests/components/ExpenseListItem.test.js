import React from 'react';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

test('should render ExpenseListItem with fixture data', () => {
  const createMockStore = configureMockStore([thunk]);
  const store = createMockStore();
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} store={store} />);
  expect(wrapper).toMatchSnapshot();
});

