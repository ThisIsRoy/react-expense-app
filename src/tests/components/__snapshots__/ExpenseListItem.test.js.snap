// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with fixture data 1`] = `
<div>
  <h3>
    gum
  </h3>
  <p>
    Amount: 
    $1.95
  </p>
  <p>
    Created: 
    December 31st, 1969
  </p>
  <Link
    replace={false}
    to="/edit/1"
  >
    <button>
      Edit
    </button>
  </Link>
</div>
`;
